File: src/Engine/Action/MoveHandler.php
<?php

declare(strict_types=1);

namespace App\Engine\Action;

use App\DTO\ActionResult;
use App\DTO\BallState;
use App\DTO\GameEvent;
use App\DTO\GameState;
use App\DTO\MatchPlayerDTO;
use App\Enum\PlayerState;
use App\Enum\SkillName;
use App\Enum\Weather;
use App\ValueObject\Position;
use App\Engine\BallResolver;
use App\Engine\DiceRollerInterface;
use App\Engine\Pathfinder;
use App\Engine\TacklezoneCalculator;

final class MoveHandler implements ActionHandlerInterface
{
    public function __construct(
        private readonly DiceRollerInterface $dice,
        private readonly TacklezoneCalculator $tzCalc,
        private readonly Pathfinder $pathfinder,
        private readonly BallResolver $ballResolver,
    ) {
    }

    /**
     * @param array<string, mixed> $params
     */
    public function resolve(GameState $state, array $params): ActionResult
    {
        $playerId = (int) $params['playerId'];
        $targetX = (int) $params['x'];
        $targetY = (int) $params['y'];
        $destination = new Position($targetX, $targetY);

        $player = $state->getPlayer($playerId);
        if ($player === null) {
            throw new \InvalidArgumentException("Player {$playerId} not found");
        }

        $events = [];

        // Handle PRONE player stand-up
        if ($player->getState() === PlayerState::PRONE) {
            $standUpResult = $this->resolveStandUp($state, $player);
            $events = array_merge($events, $standUpResult['events']);
            $state = $standUpResult['state'];

            if (!$standUpResult['success']) {
                // Failed stand-up: player stays prone, action done (not a turnover)
                return ActionResult::success($state, $events);
            }

            $player = $state->getPlayer($playerId);
            if ($player === null) {
                throw new \InvalidArgumentException("Player {$playerId} not found after stand-up");
            }

            // Stand in place only
            if ($player->getPosition() !== null && $player->getPosition()->equals($destination)) {
                $state = $state->withPlayer($player->withHasMoved(true));
                return ActionResult::success($state, $events);
            }
        }

        $path = $this->pathfinder->findPathTo($state, $player, $destination);
        if ($path === null) {
            throw new \InvalidArgumentException("No valid path to ({$targetX},{$targetY})");
        }

        $currentState = $state;
        $currentPlayer = $player;

        // Track team reroll availability
        $activeSide = $currentState->getActiveTeam();
        $teamRerollUsed = false;

        foreach ($path->getSteps() as $step) {
            $from = $currentPlayer->getPosition();
            $to = $step->getPosition();

            // Leap: special AG roll, skip dodge and normal movement
            if ($step->isLeap()) {
                $leapTarget = $step->getDodgeTarget();
                // Very Long Legs: -1 to leap target
                if ($currentPlayer->hasSkill(SkillName::VeryLongLegs)) {
                    $leapTarget = max(2, $leapTarget - 1);
                }
                $roll = $this->dice->rollD6();
                $success = $roll >= $leapTarget;
                $events[] = GameEvent::leap($playerId, $leapTarget, $roll, $success);

                if (!$success) {
                    // Failed leap: same as failed dodge
                    $events[] = GameEvent::playerFell($playerId);
                    $events[] = GameEvent::turnover('Failed leap');

                    $fallenPlayer = $currentPlayer
                        ->withState(PlayerState::PRONE)
                        ->withPosition($to)
                        ->withHasMoved(true)
                        ->withHasActed(true)
                        ->withMovementRemaining(0);
                    $currentState = $currentState->withPlayer($fallenPlayer);

                    [$currentState, $events] = $this->ballResolver->handleBallOnPlayerDown($currentState, $fallenPlayer, $events);

                    return ActionResult::turnover(
                        $currentState->withTurnoverPending(true),
                        $events,
                    );
                }

                // Successful leap: move player
                $fromStr = $from !== null ? (string) $from : '?';
                $events[] = GameEvent::playerMove($playerId, $fromStr, (string) $to);

                $currentPlayer = $currentPlayer
                    ->withPosition($to)
                    ->withMovementRemaining($currentPlayer->getMovementRemaining() - 2);
                $currentState = $currentState->withPlayer($currentPlayer);

                if ($currentState->getBall()->getCarrierId() === $playerId) {
                    $currentState = $currentState->withBall(BallState::carried($to, $playerId));
                }

                // Check pickup on landing
                $ball = $currentState->getBall();
                if (!$ball->isHeld() && $ball->getPosition() !== null && $ball->getPosition()->equals($to)) {
                    $currentPlayer = $currentState->getPlayer($playerId) ?? $currentPlayer;
                    $pickupRerollAvailable = !$teamRerollUsed && $currentState->getTeamState($activeSide)->canUseReroll();
                    $pickupResult = $this->ballResolver->resolvePickup($currentState, $currentPlayer, $pickupRerollAvailable);
                    $events = array_merge($events, $pickupResult['events']);
                    $currentState = $pickupResult['state'];

                    if ($pickupResult['teamRerollUsed']) {
                        $teamRerollUsed = true;
                        $currentState = $currentState->withTeamState(
                            $activeSide,
                            $currentState->getTeamState($activeSide)->withRerollUsed(),
                        );
                    }

                    if (!$pickupResult['success']) {
                        $events[] = GameEvent::turnover('Failed pickup');
                        $fallenPlayer = ($currentState->getPlayer($playerId) ?? $currentPlayer)
                            ->withHasMoved(true)
                            ->withHasActed(true);
                        $currentState = $currentState->withPlayer($fallenPlayer);

                        return ActionResult::turnover(
                            $currentState->withTurnoverPending(true),
                            $events,
                        );
                    }

                    $currentPlayer = $currentState->getPlayer($playerId) ?? $currentPlayer;
                }

                continue; // skip normal dodge/GFI for this step
            }

            // Tentacles check before dodge
            if ($step->requiresDodge() && $from !== null) {
                foreach ($this->tzCalc->getMarkingPlayers($currentState, $from, $currentPlayer->getTeamSide()) as $marker) {
                    if (!$marker->hasSkill(SkillName::Tentacles)) {
                        continue;
                    }
                    $moverRoll = $this->dice->rollD6();
                    $tentRoll = $this->dice->rollD6();
                    $escaped = ($moverRoll + $currentPlayer->getStats()->getStrength()) > ($tentRoll + $marker->getStats()->getStrength());
                    $events[] = GameEvent::tentacles($playerId, $marker->getId(), $moverRoll, $tentRoll, $escaped);
                    if (!$escaped) {
                        // Movement ends, NOT a turnover
                        $currentPlayer = $currentPlayer->withHasMoved(true);
                        $currentState = $currentState->withPlayer($currentPlayer);
                        return ActionResult::success($currentState, $events);
                    }
                }
            }

            // Dodge roll
            if ($step->requiresDodge()) {
                $roll = $this->dice->rollD6();
                $target = $step->getDodgeTarget();
                $success = $roll >= $target;
                $events[] = GameEvent::dodgeAttempt($playerId, $target, $roll, $success);

                if (!$success) {
                    // Try Dodge skill reroll (negated by adjacent enemy with Tackle)
                    $skillRerollUsed = false;
                    if ($currentPlayer->hasSkill(SkillName::Dodge)) {
                        $tackleNegatesDodge = false;
                        if ($from !== null) {
                            foreach ($this->tzCalc->getMarkingPlayers($currentState, $from, $currentPlayer->getTeamSide()) as $marker) {
                                if ($marker->hasSkill(SkillName::Tackle)) {
                                    $tackleNegatesDodge = true;
                                    break;
                                }
                            }
                        }

                        if (!$tackleNegatesDodge) {
                            $skillRerollUsed = true;
                            $roll = $this->dice->rollD6();
                            $success = $roll >= $target;
                            $events[] = GameEvent::rerollUsed($playerId, 'Dodge');
                            $events[] = GameEvent::dodgeAttempt($playerId, $target, $roll, $success);
                        }
                    }

                    // Pro reroll (after skill reroll, before team reroll)
                    if (!$success && $currentPlayer->hasSkill(SkillName::Pro) && !$currentPlayer->isProUsedThisTurn()) {
                        $proRoll = $this->dice->rollD6();
                        $currentPlayer = $currentPlayer->withProUsedThisTurn(true);
                        $currentState = $currentState->withPlayer($currentPlayer);
                        if ($proRoll >= 4) {
                            $roll = $this->dice->rollD6();
                            $success = $roll >= $target;
                            $events[] = GameEvent::proReroll($playerId, $proRoll, true, $roll);
                            $events[] = GameEvent::dodgeAttempt($playerId, $target, $roll, $success);
                        } else {
                            $events[] = GameEvent::proReroll($playerId, $proRoll, false, null);
                        }
                    }

                    // Try team reroll (only if no skill reroll was used)
                    if (!$success && !$skillRerollUsed && !$teamRerollUsed && $currentState->getTeamState($activeSide)->canUseReroll()) {
                        $teamRerollUsed = true;
                        $currentState = $currentState->withTeamState(
                            $activeSide,
                            $currentState->getTeamState($activeSide)->withRerollUsed(),
                        );
                        $lonerBlocked = false;
                        if ($currentPlayer->hasSkill(SkillName::Loner)) {
                            $lonerRoll = $this->dice->rollD6();
                            $lonerBlocked = $lonerRoll < 4;
                            $events[] = GameEvent::lonerCheck($playerId, $lonerRoll, !$lonerBlocked);
                        }
                        if (!$lonerBlocked) {
                            $roll = $this->dice->rollD6();
                            $success = $roll >= $target;
                            $events[] = GameEvent::rerollUsed($playerId, 'Team Reroll');
                            $events[] = GameEvent::dodgeAttempt($playerId, $target, $roll, $success);
                        }
                    }
                }

                if (!$success) {
                    // Failed dodge - player falls, turnover
                    $events[] = GameEvent::playerFell($playerId);
                    $events[] = GameEvent::turnover('Failed dodge');

                    $fallenPlayer = $currentPlayer
                        ->withState(PlayerState::PRONE)
                        ->withHasMoved(true)
                        ->withHasActed(true)
                        ->withMovementRemaining(0);
                    $currentState = $currentState->withPlayer($fallenPlayer);

                    // Drop ball if carrier
                    [$currentState, $events] = $this->ballResolver->handleBallOnPlayerDown($currentState, $fallenPlayer, $events);

                    return ActionResult::turnover(
                        $currentState->withTurnoverPending(true),
                        $events,
                    );
                }

                // Diving Tackle: after successful dodge, DT player goes prone
                if ($from !== null) {
                    foreach ($this->tzCalc->getMarkingPlayers($currentState, $from, $currentPlayer->getTeamSide()) as $marker) {
                        if (!$marker->hasSkill(SkillName::DivingTackle)) {
                            continue;
                        }
                        $dtPlayer = $marker->withState(PlayerState::PRONE);
                        $currentState = $currentState->withPlayer($dtPlayer);
                        $events[] = GameEvent::divingTackle($marker->getId());
                        [$currentState, $events] = $this->ballResolver->handleBallOnPlayerDown($currentState, $dtPlayer, $events);
                        break;
                    }
                }

                // Shadowing: opponent with Shadowing may follow after successful dodge
                if ($from !== null) {
                    foreach ($this->tzCalc->getMarkingPlayers($currentState, $from, $currentPlayer->getTeamSide()) as $marker) {
                        if (!$marker->hasSkill(SkillName::Shadowing)) {
                            continue;
                        }
                        $shadowRoll = $this->dice->rollD6();
                        $shadowMA = $marker->getStats()->getMovement();
                        $moverMA = $currentPlayer->getStats()->getMovement();
                        $followed = ($shadowRoll + $shadowMA - $moverMA) >= 6;
                        $events[] = GameEvent::shadowing($marker->getId(), $currentPlayer->getId(), $shadowRoll, $followed);
                        if ($followed) {
                            $marker = $marker->withPosition($from);
                            $currentState = $currentState->withPlayer($marker);
                        }
                        break; // only one Shadowing attempt per dodge step
                    }
                }
            }

            // GFI roll
            if ($step->isGfi()) {
                $gfiThreshold = $currentState->getWeather() === Weather::BLIZZARD ? 3 : 2;
                $roll = $this->dice->rollD6();
                $success = $roll >= $gfiThreshold;
                $events[] = GameEvent::gfiAttempt($playerId, $roll, $success);

                // Sure Feet: reroll failed GFI
                $skillRerollUsedGfi = false;
                if (!$success && $currentPlayer->hasSkill(SkillName::SureFeet)) {
                    $skillRerollUsedGfi = true;
                    $roll = $this->dice->rollD6();
                    $success = $roll >= $gfiThreshold;
                    $events[] = GameEvent::rerollUsed($playerId, 'Sure Feet');
                    $events[] = GameEvent::gfiAttempt($playerId, $roll, $success);
                }

                // Pro reroll (after skill reroll, before team reroll)
                if (!$success && $currentPlayer->hasSkill(SkillName::Pro) && !$currentPlayer->isProUsedThisTurn()) {
                    $proRoll = $this->dice->rollD6();
                    $currentPlayer = $currentPlayer->withProUsedThisTurn(true);
                    $currentState = $currentState->withPlayer($currentPlayer);
                    if ($proRoll >= 4) {
                        $roll = $this->dice->rollD6();
                        $success = $roll >= $gfiThreshold;
                        $events[] = GameEvent::proReroll($playerId, $proRoll, true, $roll);
                        $events[] = GameEvent::gfiAttempt($playerId, $roll, $success);
                    } else {
                        $events[] = GameEvent::proReroll($playerId, $proRoll, false, null);
                    }
                }

                // Try team reroll (only if no skill reroll was used)
                if (!$success && !$skillRerollUsedGfi && !$teamRerollUsed && $currentState->getTeamState($activeSide)->canUseReroll()) {
                    $teamRerollUsed = true;
                    $currentState = $currentState->withTeamState(
                        $activeSide,
                        $currentState->getTeamState($activeSide)->withRerollUsed(),
                    );
                    $lonerBlocked = false;
                    if ($currentPlayer->hasSkill(SkillName::Loner)) {
                        $lonerRoll = $this->dice->rollD6();
                        $lonerBlocked = $lonerRoll < 4;
                        $events[] = GameEvent::lonerCheck($playerId, $lonerRoll, !$lonerBlocked);
                    }
                    if (!$lonerBlocked) {
                        $roll = $this->dice->rollD6();
                        $success = $roll >= $gfiThreshold;
                        $events[] = GameEvent::rerollUsed($playerId, 'Team Reroll');
                        $events[] = GameEvent::gfiAttempt($playerId, $roll, $success);
                    }
                }

                if (!$success) {
                    // Failed GFI - player falls, turnover
                    $events[] = GameEvent::playerFell($playerId);
                    $events[] = GameEvent::turnover('Failed Going For It');

                    $fallenPlayer = $currentPlayer
                        ->withState(PlayerState::PRONE)
                        ->withPosition($to) // falls at destination
                        ->withHasMoved(true)
                        ->withHasActed(true)
                        ->withMovementRemaining(0);
                    $currentState = $currentState->withPlayer($fallenPlayer);

                    // Drop ball if carrier
                    [$currentState, $events] = $this->ballResolver->handleBallOnPlayerDown($currentState, $fallenPlayer, $events);

                    return ActionResult::turnover(
                        $currentState->withTurnoverPending(true),
                        $events,
                    );
                }
            }

            // Move succeeded for this step
            $fromStr = $from !== null ? (string) $from : '?';
            $events[] = GameEvent::playerMove($playerId, $fromStr, (string) $to);

            $currentPlayer = $currentPlayer
                ->withPosition($to)
                ->withMovementRemaining($currentPlayer->getMovementRemaining() - 1);
            $currentState = $currentState->withPlayer($currentPlayer);

            // Move ball with carrier
            if ($currentState->getBall()->getCarrierId() === $playerId) {
                $currentState = $currentState->withBall(BallState::carried($to, $playerId));
            }

            // Check if player moved onto a loose ball
            $ball = $currentState->getBall();
            if (!$ball->isHeld() && $ball->getPosition() !== null && $ball->getPosition()->equals($to)) {
                $currentPlayer = $currentState->getPlayer($playerId) ?? $currentPlayer;
                $pickupRerollAvailable = !$teamRerollUsed && $currentState->getTeamState($activeSide)->canUseReroll();
                $pickupResult = $this->ballResolver->resolvePickup($currentState, $currentPlayer, $pickupRerollAvailable);
                $events = array_merge($events, $pickupResult['events']);
                $currentState = $pickupResult['state'];

                if ($pickupResult['teamRerollUsed']) {
                    $teamRerollUsed = true;
                    $currentState = $currentState->withTeamState(
                        $activeSide,
                        $currentState->getTeamState($activeSide)->withRerollUsed(),
                    );
                }

                if (!$pickupResult['success']) {
                    $events[] = GameEvent::turnover('Failed pickup');
                    $fallenPlayer = ($currentState->getPlayer($playerId) ?? $currentPlayer)
                        ->withHasMoved(true)
                        ->withHasActed(true);
                    $currentState = $currentState->withPlayer($fallenPlayer);

                    return ActionResult::turnover(
                        $currentState->withTurnoverPending(true),
                        $events,
                    );
                }

                $currentPlayer = $currentState->getPlayer($playerId) ?? $currentPlayer;
            }
        }

        // Mark player as moved
        $movedPlayer = $currentPlayer->withHasMoved(true);
        $currentState = $currentState->withPlayer($movedPlayer);

        return ActionResult::success($currentState, $events);
    }

    /**
     * Resolve stand-up for a prone player.
     * Jump Up: free stand-up (0 MA cost).
     * MA >= 3: automatic, costs 3 MA. MA < 3: roll 4+ required.
     *
     * @return array{state: GameState, events: list<GameEvent>, success: bool}
     */
    private function resolveStandUp(GameState $state, MatchPlayerDTO $player): array
    {
        $playerId = $player->getId();

        // Jump Up: stand up for free (0 MA cost)
        if ($player->hasSkill(SkillName::JumpUp)) {
            $events = [GameEvent::standUp($playerId)];
            $player = $player->withState(PlayerState::STANDING);
            $state = $state->withPlayer($player);
            return ['state' => $state, 'events' => $events, 'success' => true];
        }

        if ($player->getStats()->getMovement() < 3) {
            // Low MA: roll 4+ to stand
            $roll = $this->dice->rollD6();
            $success = $roll >= 4;
            $events = [GameEvent::standUp($playerId, $roll, $success)];

            if (!$success) {
                // Failed: player stays prone, action done
                $player = $player->withHasMoved(true)->withHasActed(true);
                $state = $state->withPlayer($player);
                return ['state' => $state, 'events' => $events, 'success' => false];
            }

            // Success: stand up with 0 MA remaining
            $player = $player
                ->withState(PlayerState::STANDING)
                ->withMovementRemaining(0);
            $state = $state->withPlayer($player);
            return ['state' => $state, 'events' => $events, 'success' => true];
        }

        // Normal stand-up: costs 3 MA
        $events = [GameEvent::standUp($playerId)];
        $player = $player
            ->withState(PlayerState::STANDING)
            ->withMovementRemaining($player->getMovementRemaining() - 3);
        $state = $state->withPlayer($player);

        return ['state' => $state, 'events' => $events, 'success' => true];
    }
}

File: scratchpad/blitz_floor_20260723/engine_patched/src/move_handler.cpp
#include "bb/move_handler.h"
#include "bb/helpers.h"
#include "bb/injury.h"
#include "bb/ball_handler.h"
#include <algorithm>

namespace bb {

namespace {

// Check Tentacles: adjacent enemies with Tentacles contest the move
// Returns true if player is caught (movement ends)
bool checkTentacles(GameState& state, int playerId, Position from,
                    DiceRollerBase& dice, std::vector<GameEvent>* events) {
    Player& mover = state.getPlayer(playerId);
    auto adj = from.getAdjacent();

    for (auto& pos : adj) {
        if (!pos.isOnPitch()) continue;
        const Player* opp = state.getPlayerAtPosition(pos);
        if (!opp || opp->teamSide == mover.teamSide) continue;
        if (!canAct(opp->state) || opp->lostTacklezones) continue;
        if (!opp->hasSkill(SkillName::Tentacles)) continue;

        // Contest: D6+moverST vs D6+tentaclesST, strictly greater to escape
        int moverRoll = dice.rollD6();
        int tentRoll = dice.rollD6();
        bool escaped = (moverRoll + mover.stats.strength) > (tentRoll + opp->stats.strength);

        emitEvent(events, {GameEvent::Type::SKILL_USED, opp->id, playerId, {}, {},
                          static_cast<int>(SkillName::Tentacles), !escaped});

        if (!escaped) {
            // Caught: movement ends, player stays at from
            mover.hasMoved = true;
            return true;
        }
        // Only one Tentacles check per dodge step
        break;
    }
    return false;
}

// Check Shadowing: after successful dodge, adjacent enemy with Shadowing may follow
void checkShadowing(GameState& state, int playerId, Position from,
                    DiceRollerBase& dice, std::vector<GameEvent>* events) {
    Player& mover = state.getPlayer(playerId);
    auto adj = from.getAdjacent();

    for (auto& pos : adj) {
        if (!pos.isOnPitch()) continue;
        Player* opp = state.getPlayerAtPosition(pos);
        if (!opp || opp->teamSide == mover.teamSide) continue;
        if (!canAct(opp->state) || opp->lostTacklezones) continue;
        if (!opp->hasSkill(SkillName::Shadowing)) continue;

        // Roll: D6 + shadowMA - moverMA. If >= 6, follower moves to vacated square
        int roll = dice.rollD6();
        int total = roll + opp->stats.movement - mover.stats.movement;
        bool follows = (total >= 6);

        emitEvent(events, {GameEvent::Type::SKILL_USED, opp->id, playerId, opp->position, from,
                          static_cast<int>(SkillName::Shadowing), follows});

        if (follows) {
            // Check that vacated square is empty (should be, since we just left)
            if (!state.getPlayerAtPosition(from)) {
                opp->position = from;
            }
        }
        // Only one Shadowing attempt per dodge step
        break;
    }
}

} // anonymous namespace

ActionResult resolveMoveStep(GameState& state, int playerId, Position to,
                             DiceRollerBase& dice, std::vector<GameEvent>* events) {
    Player& player = state.getPlayer(playerId);
    Position from = player.position;

    // Validate: adjacent and on-pitch
    if (from.distanceTo(to) != 1 || !to.isOnPitch()) {
        return ActionResult::fail();
    }

    // Validate: destination not occupied
    if (state.getPlayerAtPosition(to) != nullptr) {
        return ActionResult::fail();
    }

    // Check if leaving a tackle zone → dodge required
    bool needsDodge = countTacklezones(state, from, player.teamSide) > 0;

    // Check Tentacles before dodge (if leaving TZ)
    if (needsDodge) {
        if (checkTentacles(state, playerId, from, dice, events)) {
            return ActionResult::ok();  // Caught, movement ends, not turnover
        }
    }

    // Decrement movement
    player.movementRemaining--;
    player.hasMoved = true;

    // Check GFI
    bool needsGfi = false;
    if (player.movementRemaining < 0) {
        // Allow up to -2 GFI squares (or -3 with Sprint)
        int maxGfi = player.hasSkill(SkillName::Sprint) ? -3 : -2;
        if (player.movementRemaining < maxGfi) {
            player.movementRemaining++; // undo
            return ActionResult::fail();
        }
        needsGfi = true;
    }

    // Perform dodge roll if needed
    if (needsDodge) {
        int target = calculateDodgeTarget(state, player, to, from);

        // Check if Tackle negates Dodge reroll
        bool tackleNegates = false;
        auto srcAdj = from.getAdjacent();
        for (auto& apos : srcAdj) {
            if (!apos.isOnPitch()) continue;
            const Player* opp = state.getPlayerAtPosition(apos);
            if (opp && opp->teamSide != player.teamSide &&
                exertsTacklezone(opp->state) && !opp->lostTacklezones &&
                opp->hasSkill(SkillName::Tackle)) {
                tackleNegates = true;
                break;
            }
        }

        bool dodgeOk = attemptRoll(state, playerId, dice, target,
                                    SkillName::Dodge, tackleNegates, true, events);

        emitEvent(events, {GameEvent::Type::DODGE, playerId, -1, from, to,
                          target, dodgeOk});

        if (!dodgeOk) {
            // Failed dodge: player falls at destination
            player.position = to;
            player.state = PlayerState::PRONE;
            player.hasActed = true;

            InjuryContext ctx;
            resolveArmourAndInjury(state, playerId, dice, ctx, events);
            handleBallOnPlayerDown(state, playerId, dice, events);

            return ActionResult::turnovr();
        }
    }

    // Perform GFI roll if needed
    if (needsGfi) {
        int gfiTarget = (state.weather == Weather::BLIZZARD) ? 3 : 2;

        bool gfiOk = attemptRoll(state, playerId, dice, gfiTarget,
                                  SkillName::SureFeet, false, true, events);

        emitEvent(events, {GameEvent::Type::GFI, playerId, -1, from, to,
                          gfiTarget, gfiOk});

        if (!gfiOk) {
            // Failed GFI: player falls at destination
            player.position = to;
            player.state = PlayerState::PRONE;
            player.hasActed = true;

            InjuryContext ctx;
            resolveArmourAndInjury(state, playerId, dice, ctx, events);
            handleBallOnPlayerDown(state, playerId, dice, events);

            return ActionResult::turnovr();
        }
    }

    // Move player
    player.position = to;

    // Update ball position if carrier
    if (state.ball.isHeld && state.ball.carrierId == playerId) {
        state.ball.position = to;
    }

    emitEvent(events, {GameEvent::Type::PLAYER_MOVE, playerId, -1, from, to, 0, true});

    // Shadowing: after successful dodge, enemy may follow
    if (needsDodge) {
        checkShadowing(state, playerId, from, dice, events);
    }

    // Pickup ball if on ground at destination
    if (!state.ball.isHeld && state.ball.position == to) {
        bool pickupOk = resolvePickup(state, playerId, dice, events);
        if (!pickupOk) {
            // Failed pickup — turnover
            player.hasActed = true;
            return ActionResult::turnovr();
        }
    }

    return ActionResult::ok();
}

ActionResult resolveLeap(GameState& state, int playerId, Position to,
                         DiceRollerBase& dice, std::vector<GameEvent>* events) {
    Player& player = state.getPlayer(playerId);
    Position from = player.position;

    // Validate: within distance 2, on pitch, unoccupied
    int dist = from.distanceTo(to);
    if (dist < 1 || dist > 2 || !to.isOnPitch()) {
        return ActionResult::fail();
    }
    if (state.getPlayerAtPosition(to) != nullptr) {
        return ActionResult::fail();
    }

    // Leap costs 2 MA
    player.movementRemaining -= 2;
    player.hasMoved = true;

    bool needsGfi = false;
    if (player.movementRemaining < 0) {
        int maxGfi = player.hasSkill(SkillName::Sprint) ? -3 : -2;
        if (player.movementRemaining < maxGfi) {
            player.movementRemaining += 2; // undo
            return ActionResult::fail();
        }
        needsGfi = true;
    }

    // Leap agility check: target = max(2, min(6, 7-AG+TZ_at_dest))
    int target = 7 - player.stats.agility;
    target += countTacklezones(state, to, player.teamSide);
    if (player.hasSkill(SkillName::VeryLongLegs)) target -= 1;
    target = std::clamp(target, 2, 6);

    bool leapOk = attemptRoll(state, playerId, dice, target,
                               SkillName::SKILL_COUNT, false, true, events);

    emitEvent(events, {GameEvent::Type::DODGE, playerId, -1, from, to,
                      target, leapOk});

    if (!leapOk) {
        // Failed leap: player prone at destination, armor+injury, turnover
        player.position = to;
        player.state = PlayerState::PRONE;
        player.hasActed = true;

        InjuryContext ctx;
        resolveArmourAndInjury(state, playerId, dice, ctx, events);
        handleBallOnPlayerDown(state, playerId, dice, events);

        return ActionResult::turnovr();
    }

    // GFI if needed
    if (needsGfi) {
        int gfiTarget = (state.weather == Weather::BLIZZARD) ? 3 : 2;
        bool gfiOk = attemptRoll(state, playerId, dice, gfiTarget,
                                  SkillName::SureFeet, false, true, events);
        if (!gfiOk) {
            player.position = to;
            player.state = PlayerState::PRONE;
            player.hasActed = true;

            InjuryContext ctx;
            resolveArmourAndInjury(state, playerId, dice, ctx, events);
            handleBallOnPlayerDown(state, playerId, dice, events);

            return ActionResult::turnovr();
        }
    }

    // Move player
    player.position = to;

    if (state.ball.isHeld && state.ball.carrierId == playerId) {
        state.ball.position = to;
    }

    emitEvent(events, {GameEvent::Type::PLAYER_MOVE, playerId, -1, from, to, 0, true});

    // Pickup ball if on ground at destination
    if (!state.ball.isHeld && state.ball.position == to) {
        bool pickupOk = resolvePickup(state, playerId, dice, events);
        if (!pickupOk) {
            player.hasActed = true;
            return ActionResult::turnovr();
        }
    }

    return ActionResult::ok();
}

ActionResult resolveStandUp(GameState& state, int playerId, DiceRollerBase& dice,
                            std::vector<GameEvent>* events) {
    Player& player = state.getPlayer(playerId);

    if (player.state != PlayerState::PRONE) {
        return ActionResult::fail();
    }

    if (player.hasSkill(SkillName::JumpUp)) {
        // Free stand up
        player.state = PlayerState::STANDING;
        return ActionResult::ok();
    }

    // Costs 3 MA
    if (player.movementRemaining < 3) {
        return ActionResult::fail();
    }

    player.movementRemaining -= 3;
    player.state = PlayerState::STANDING;
    return ActionResult::ok();
}

} // namespace bb
